internals/secrethub/tpl/v2.go

```diff
@@ -24,95 +37,418 @@ func NewV2Parser() Parser {
-func (t secretTemplateV2) InjectSecrets(secrets map[string]string) (string, error) {
-	keys := make(map[string]string, len(t.secrets))
-	for k, secretpath := range t.secrets {
-		v, ok := secrets[secretpath]
-		if !ok {
-			return "", fmt.Errorf("no value supplied for secret %s", secretpath)
+				err := p.readRune()
+				if err == io.EOF {
+					return nil, ErrVariableTagNotClosed(p.lineNo, p.columnNo+1)
+				}
+				if err != nil {
+					return nil, err
+				}
+				continue
+			}
+			return nil, ErrIllegalVariableCharacter(p.next, p.lineNo, p.columnNo+1)
 		}
-		keys[k] = v
 	}
-	return t.template.Inject(keys)
 }
 
-// Secrets returns a list of paths to secrets that are used in the template.
-func (t secretTemplateV2) Secrets() []string {
-	set := map[string]struct{}{}
-	for _, path := range t.secrets {
-		set[path] = struct{}{}
+// parseSecret parses the contents of a secret tag up to the closing delimiter.
+// parseSecret should be called after the opening delimiter has been read. The next
+// character from the buffer should be the first character of the contents.
+//
+// when parseSecret returns, the next character in the buffer is the first character
+// after the closing delimiter of the secret tag.
+func (p *v2Parser) parseSecret() (node, error) {
+	path := []node{}
+	err := p.readRune()
+	if err == io.EOF {
+		return nil, ErrSecretTagNotClosed(p.lineNo, p.columnNo+1)
+	}
+	if err != nil {
+		return nil, err
+	}
+	for p.next == ' ' {
+		err = p.readRune()
+		if err == io.EOF {
+			return nil, ErrSecretTagNotClosed(p.lineNo, p.columnNo+1)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for {
+		err = p.readRune()
+		if err == io.EOF {
+			return nil, ErrSecretTagNotClosed(p.lineNo, p.columnNo+1)
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		switch p.current {
+		case '$':
+			switch p.next {
+			case '{':
+				err = p.readRune()
+				if err == io.EOF {
+					return nil, ErrVariableTagNotClosed(p.lineNo, p.columnNo+1)
+				}
+				if err != nil {
+					return nil, err
+				}
+				variable, err := p.parseVar()
+				if err != nil {
+					return nil, err
+				}
+				path = append(path, variable)
+
+				err = p.readRune()
+				if err == io.EOF {
+					return nil, ErrSecretTagNotClosed(p.lineNo, p.columnNo+1)
+				}
+				if err != nil {
+					return nil, err
+				}
+			default:
+				return nil, ErrIllegalSecretCharacter(p.current, p.lineNo, p.columnNo)
+			}
+		case ' ':
+			err := p.forwardToClosing([]rune("}}"))
+			if err != nil {
+				return nil, ErrIllegalSecretCharacter(p.current, p.lineNo, p.columnNo)
+			}
+			return secret{
+				path: path,
+			}, nil
+		case '}':
+			switch p.next {
+			case '}':
+				return secret{
+					path: path,
+				}, nil
+			default:
+				return nil, ErrIllegalSecretCharacter(p.current, p.lineNo, p.columnNo)
+			}
+		default:
+			if unicode.IsLetter(p.current) || unicode.IsDigit(p.current) || p.current == '_' || p.current == '-' || p.current == '.' || p.current == '/' || p.current == ':' {
```

Comment:
Consider creating a small helper function for this, e.g. `func isValidSecretPathRune(r rune) bool` and while you're at it, create his buddy too: `isValidVariableRune`

Reply:
e76bf2ff93dea5de39d1bd11fee1a631e4756589